image_id.go

```diff
@@ -0,0 +1,103 @@
+package flux
+
+import (
+	"fmt"
+	"strings"
+
+	"github.com/pkg/errors"
+)
+
+const (
+	dockerHubHost    = "index.docker.io"
+	dockerHubLibrary = "library"
+)
+
+var (
+	ErrInvalidImageID   = errors.New("invalid image ID")
+	ErrBlankImageID     = errors.Wrap(ErrInvalidImageID, "blank image name")
+	ErrMalformedImageID = errors.Wrap(ErrInvalidImageID, `expected image name as either <image>:<tag> or just <image>`)
+)
+
+// ImageID is a fully qualified name that refers to a particular Image.
+// It is in the format: host[:port]/Namespace/Image[:tag]
+// Here, we refer to the "name" == Namespace/Image
+type ImageID struct {
+	Host, Namespace, Image, Tag string
+}
+
+func ParseImageID(s string) (ImageID, error) {
+	if s == "" {
+		return ImageID{}, ErrBlankImageID
+	}
+	var img ImageID
+	parts := strings.Split(s, ":")
+	switch len(parts) {
+	case 0:
+		return ImageID{}, ErrMalformedImageID
+	case 1:
+		img.Tag = "latest"
+	case 2:
+		img.Tag = parts[1]
+		s = parts[0]
+	default:
+		return ImageID{}, ErrMalformedImageID
+	}
+	if s == "" {
+		return ImageID{}, ErrBlankImageID
+	}
+	parts = strings.Split(s, "/")
+	switch len(parts) {
+	case 1:
+		img.Host = dockerHubHost
+		img.Namespace = dockerHubLibrary
+		img.Image = parts[0]
+	case 2:
+		img.Host = dockerHubHost
+		img.Namespace = parts[0]
+		img.Image = parts[1]
+	case 3:
+		img.Host = parts[0]
+		img.Namespace = parts[1]
+		img.Image = parts[2]
+	default:
+		return ImageID{}, ErrMalformedImageID
+	}
+	return img, nil
+}
+
+// Fully qualified name
+func (i ImageID) String() string {
+	if i.Image == "" {
+		return "" // Doesn't make sense to return anything if it doesn't even have an image
+	}
+	var ta string
+	if i.Tag != "" {
+		ta = fmt.Sprintf(":%s", i.Tag)
+	}
+	return fmt.Sprintf("%s%s", i.Repository(), ta)
+}
+
+// Repository returns the short version of an image's repository (trimming if dockerhub)
+func (i ImageID) Repository() string {
```

Comment:
Usually we want one of three things:

  1. are these images equivalent? (compare canonical IDs)
  2. are these images from the same repo? (compare canonical repository name)
  3. show the repository name as given

This suggests to me that the most important methods are, say, `Equivalent(imageID)` and `SameRepository(imageID)`. I'm not sure we ever display canonical names/IDs, so those might not even need to be calculated outside of those methods.

Since we mostly want to display image names as given, we probably don't want to add defaults in when parsing (only when comparing).